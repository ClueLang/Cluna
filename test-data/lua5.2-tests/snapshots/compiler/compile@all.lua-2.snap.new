---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/all.lua
---
local version = "Lua 5.2"
if _VERSION ~= version {
    io.stderr::write("\nThis test suite is for ", version, ", not for ", _VERSION, "\nExiting tests\n");
    return 
}
_soft = false
_port = false
_no32 = false
_nomsg = false
_noposix = false
_nolonglong = false
_noformatA = false
local usertests = rawget(_G, "_U")
if usertests {
    _soft = true
    _port = true
    _no32 = true
    _nomsg = true
    _noposix = true
    _nolonglong = true
    _noformatA = true
}
if usertests {
    T = nil
}
T = rawget(_G, "T")
package.path = "?;./?.lua" .. package.path
math.randomseed(0);
collectgarbage("setstepmul", 200);
collectgarbage("setpause", 200);
print("current path:\n****" .. package.path .. "****\n");
local c = os.clock()
local collectgarbage = collectgarbage
{
    local msgs = {}
    global fn Message(m) {
        if !_nomsg {
            print(m);
            msgs[(#msgs + 1)] = string.sub(m, 3,  - 3)
        }
    }
    assert(os.setlocale("C"));
    local T, print, format, write, assert, type, unpack, floor = T, print, string.format, io.write, assert, type, table.unpack, math.floor
    local fn F(m) {
        local fn round(m) {
            m = m + 0.04999
            return m - (m % 0.1)
        }
        if m < 1000 {
            return m
        } else {
            m = m / 1000
            if m < 1000 {
                return round(m) .. "K"
            } else {
                return round(m / 1000) .. "M"
            }
        }
    }
    local showmem
    if !T {
        local max = 0
        showmem = fn () {
            local m = collectgarbage("count") * 1024
            max = (m > max) && m || max
            print(format("    ---- total memory: %s, max memory: %s ----\n", F(m), F(max)));
        }
    } else {
        showmem = fn () {
            T.checkmemory();
            local total, numblocks, maxmem = T.totalmem()
            local count = collectgarbage("count")
            print(format("\n    ---- total memory: %s (%.0fK), max use: %s,  blocks: %d\n", F(total), count, F(maxmem), numblocks));
            print(format("\t(strings:  %d, tables: %d, functions: %d, " .. "\n\tudata: %d, threads: %d)", T.totalmem("string"), T.totalmem("table"), T.totalmem("function"), T.totalmem("userdata"), T.totalmem("thread")));
        }
    }
    local fn report(n) {
        print("\n***** FILE '" .. n .. "'*****");
    }
    local olddofile = dofile
    dofile = fn (n) {
        showmem();
        report(n);
        local f = assert(loadfile(n))
        local b = string.dump(f)
        f = assert(load(b))
        return f()
    }
    dofile('main.lua');
    {
        local next, setmetatable, stderr = next, setmetatable, io.stderr
        local mt = {}
        mt.__gc = fn (o) {
            stderr::write('.');
            local n = setmetatable({}, mt)
            o = nil
            local a, b, c, d, e = nil
        }
        local n = setmetatable({}, mt)
    }
    report("gc.lua");
    local f = assert(loadfile('gc.lua'))
    f();
    collectgarbage("generational");
    dofile('db.lua');
    assert(dofile('calls.lua') == deep && deep);
    olddofile('strings.lua');
    olddofile('literals.lua');
    assert(dofile('attrib.lua') == 27);
    collectgarbage("incremental");
    olddofile('strings.lua');
    olddofile('literals.lua');
    dofile('constructs.lua');
    dofile('api.lua');
    collectgarbage("generational");
    collectgarbage("setpause", 200);
    collectgarbage("setmajorinc", 500);
    assert(dofile('locals.lua') == 5);
    dofile('constructs.lua');
    dofile('code.lua');
    if !_G._soft {
        report('big.lua');
        local f = coroutine.wrap(assert(loadfile('big.lua')))
        assert(f() == 'b');
        assert(f() == 'a');
    }
    dofile('nextvar.lua');
    dofile('pm.lua');
    dofile('api.lua');
    assert(dofile('events.lua') == 12);
    dofile('vararg.lua');
    dofile('closure.lua');
    dofile('coroutine.lua');
    dofile('goto.lua');
    dofile('errors.lua');
    dofile('math.lua');
    dofile('sort.lua');
    dofile('bitwise.lua');
    assert(dofile('verybig.lua') == 10);
    collectgarbage();
    dofile('files.lua');
    if #msgs > 0 {
        print("\ntests not performed:");
        for i = 1, #msgs {
            print(msgs[(i)]);
        }
        print();
    }
    print("final OK !!!");
    local debug = require("debug")
    debug.sethook(fn (a) {
        assert(type(a) == 'string');
    }, "cr");
    _G.showmem = showmem
}
local _G, showmem, print, format, clock, assert, open = _G, showmem, print, string.format, os.clock, assert, io.open
local fname = T && "time-debug.txt" || "time.txt"
local lasttime
if !usertests {
    local f = io.open(fname)
    if f {
        lasttime = assert(tonumber(f::read('*a')))
        f::close();
    } else {
        lasttime = nil
    }
}
print('cleaning all!!!!');
for n with pairs(_G) {
    if !({
        ___Glob = 1, 
        tostring = 1
    })[(n)] {
        _G[(n)] = nil
    }
}
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
collectgarbage();
showmem();
local time = clock() - c
print(format("\n\ntotal time: %.2f\n", time));
if !usertests {
    lasttime = lasttime || time
    local diff = (time - lasttime) / time
    local tolerance = 0.05
    assert(diff < tolerance && diff >  - tolerance);
    assert(open(fname, "w"))::write(time)::close();
}
