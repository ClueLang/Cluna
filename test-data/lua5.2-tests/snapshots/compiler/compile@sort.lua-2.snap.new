---
source: src/main.rs
assertion_line: 118
expression: compiled
input_file: test-data/lua5.2-tests/sort.lua
---
print("testing (parts of) table library");
print("testing unpack");
local unpack = table.unpack
local x, y, z, a, n
a = {}
lim = 2000
for i = 1, lim {
    a[(i)] = i
}
assert(select(lim, unpack(a)) == lim && select('#', unpack(a)) == lim);
x = unpack(a)
assert(x == 1);
x = {
    unpack(a)
}
assert(#x == lim && x[(1)] == 1 && x[(lim)] == lim);
x = {
    unpack(a, lim - 2)
}
assert(#x == 3 && x[(1)] == lim - 2 && x[(3)] == lim);
x = {
    unpack(a, 10, 6)
}
assert(next(x) == nil);
x = {
    unpack(a, 11, 10)
}
assert(next(x) == nil);
x, y = unpack(a, 10, 10)
assert(x == 10 && y == nil);
x, y, z = unpack(a, 10, 11)
assert(x == 10 && y == 11 && z == nil);
a, x = unpack({
    1
})
assert(a == 1 && x == nil);
a, x = unpack({
    1, 
    2
}, 1, 1)
assert(a == 1 && x == nil);
if !_no32 {
    assert(!pcall(unpack, {}, 0, 2 ^ 31 - 1));
    assert(!pcall(unpack, {}, 1, 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31), 2 ^ 31 - 1));
    assert(!pcall(unpack, {},  - (2 ^ 31 - 1), 2 ^ 31 - 1));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 0));
    assert(pcall(unpack, {}, 2 ^ 31 - 1, 1));
    pcall(unpack, {}, 1, 2 ^ 31);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 1, 2 ^ 31 - 1)
    assert(a == 20 && b == nil);
    a, b = unpack({
        2 ^ 31 - 1 = 20
    }, 2 ^ 31 - 2, 2 ^ 31 - 1)
    assert(a == nil && b == 20);
}
print("testing pack");
a = table.pack()
assert(a[(1)] == nil && a.n == 0);
a = table.pack(table)
assert(a[(1)] == table && a.n == 1);
a = table.pack(nil, nil, nil, nil)
assert(a[(1)] == nil && a.n == 4);
print("testing sort");
local fn check(t) {
    local fn f(a, b) {
        assert(a && b);
        return true
    }
    local s, e = pcall(table.sort, t, f)
    assert(!s && e::find("invalid order function"));
}
check({
    1, 
    2, 
    3, 
    4
});
check({
    1, 
    2, 
    3, 
    4, 
    5
});
check({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
global fn check(a, f) {
    f = f || fn (x, y) {
        return x < y
    }
    for n = #a, 2,  - 1 {
        assert(!f(a[(n)], a[(n - 1)]));
    }
}
a = {
    "Jan", 
    "Feb", 
    "Mar", 
    "Apr", 
    "May", 
    "Jun", 
    "Jul", 
    "Aug", 
    "Sep", 
    "Oct", 
    "Nov", 
    "Dec"
}
table.sort(a);
check(a);
global fn perm(s, n) {
    n = n || #s
    if n == 1 {
        local t = {
            unpack(s)
        }
        table.sort(t);
        check(t);
    } else {
        for i = 1, n {
            s[(i)], s[(n)] = s[(n)], s[(i)]
            perm(s, n - 1);
            s[(i)], s[(n)] = s[(n)], s[(i)]
        }
    }
}
perm({});
perm({
    1
});
perm({
    1, 
    2
});
perm({
    1, 
    2, 
    3
});
perm({
    1, 
    2, 
    3, 
    4
});
perm({
    2, 
    2, 
    3, 
    4
});
perm({
    1, 
    2, 
    3, 
    4, 
    5
});
perm({
    1, 
    2, 
    3, 
    3, 
    5
});
perm({
    1, 
    2, 
    3, 
    4, 
    5, 
    6
});
perm({
    2, 
    2, 
    3, 
    3, 
    5, 
    6
});
limit = 30000
if _soft {
    limit = 5000
}
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
local x = os.clock()
table.sort(a);
print(string.format("Sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
x = os.clock()
table.sort(a);
print(string.format("Re-sorting %d elements in %.2f sec.", limit, os.clock() - x));
check(a);
a = {}
for i = 1, limit {
    a[(i)] = math.random()
}
x = os.clock()
i = 0
table.sort(a, fn (x, y) {
    i = i + 1
    return y < x
});
print(string.format("Invert-sorting other %d elements in %.2f sec., with %i comparisons", limit, os.clock() - x, i));
check(a, fn (x, y) {
    return y < x
});
table.sort({});
for i = 1, limit {
    a[(i)] = false
}
x = os.clock()
table.sort(a, fn (x, y) {
    return nil
});
print(string.format("Sorting %d equal elements in %.2f sec.", limit, os.clock() - x));
check(a, fn (x, y) {
    return nil
});
for i, v with pairs(a) {
    assert(!v || i == 'n' && v == limit);
}
A = {
    "Ã¡lo", 
    "\0first :-)", 
    "alo", 
    "then this one", 
    "45", 
    "and a new"
}
table.sort(A);
check(A);
table.sort(A, fn (x, y) {
    load(string.format("A[%q] = ''", x))();
    collectgarbage();
    return x < y
});
tt = {
    __lt = fn (a, b) {
        return a.val < b.val
    }
}
a = {}
for i = 1, 10 {
    a[(i)] = {
        val = math.random(100)
    }
    setmetatable(a[(i)], tt);
}
table.sort(a);
check(a, tt.__lt);
check(a);
print("OK");
