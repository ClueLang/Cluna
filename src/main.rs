#![allow(non_snake_case)]
#![allow(non_upper_case_globals)]

macro_rules! check {
    ($tocheck: expr) => {
        match $tocheck {
            Ok(t) => t,
            Err(e) => return Err(e.to_string()),
        }
    };
}

macro_rules! arg {
    ($name: expr) => {
        unsafe { $name }
    };
}

// mod compiler;
mod compiler;
mod parser;
mod scanner;

use clap::Parser;
use compiler::compile_tokens;
use parser::*;
use scanner::*;
use std::{fs, fs::File, io::prelude::*, path::Path, time::Instant};

pub static mut ENV_TOKENS: bool = false;
pub static mut ENV_STRUCT: bool = false;
pub static mut ENV_OUTPUT: bool = false;
pub static mut ENV_DONTSAVE: bool = false;
pub static mut ENV_PATHISCODE: bool = false;
pub static mut ENV_NOMULTILINE: bool = false;

#[derive(Parser)]
#[clap(about, version, long_about = None)]
struct Cli {
    /// The path to the directory where the *.lua files are located.
    /// Every directory inside the given directory will be checked too.
    /// If the path points to a single *.lua file, only that file will be compiled.
    #[clap(required_unless_present = "license")]
    path: Option<String>,

    /// The name the output file will have
    #[clap(default_value = "main", value_name = "OUTPUT FILE NAME")]
    outputname: String,

    /// Print license information
    #[clap(short = 'L', long, display_order = 1000)]
    license: bool,

    /// Print list of detected tokens in compiled files
    #[clap(long)]
    tokens: bool,

    /// Print syntax structure of the tokens of the compiled files
    #[clap(long)]
    r#struct: bool,

    /// Print output Clue code in the console
    #[clap(short, long)]
    output: bool,

    /// Don't save compiled code
    #[clap(short = 'D', long)]
    dontsave: bool,

    /// Treat PATH not as a path but as clue code
    #[clap(short, long)]
    pathiscode: bool,

    /// Don't use multiline strings in the output
    #[clap(long)]
    nomultiline: bool,
}

fn compile_code(code: String, name: String, scope: usize) -> Result<String, String> {
    let time = Instant::now();
    let (tokens, _comments): (Vec<Token>, Vec<Comment>) = scan_code(code, name.clone())?;
    if arg!(ENV_TOKENS) {
        println!("Scanned tokens of file \"{}\":\n{:#?}", name, tokens);
    }
    let ctokens = parse_tokens(tokens, name.clone())?;
    if arg!(ENV_STRUCT) {
        println!("Parsed structure of file \"{}\":\n{:#?}", name, ctokens);
    }

    let code = compile_tokens(scope, ctokens);
    if arg!(ENV_OUTPUT) {
        println!("Compiled Lua code of file \"{}\":\n{}", name, code);
    }
    println!(
        "Compiled file \"{}\" in {} seconds!",
        name,
        time.elapsed().as_secs_f32()
    );
    Ok(code)
}

fn compile_file(path: &Path, name: String, scope: usize) -> Result<String, String> {
    let mut code: String = String::new();
    check!(check!(File::open(path)).read_to_string(&mut code));
    Ok(compile_code(code, name, scope)?)
}

fn compile_folder(path: &Path, rpath: String) -> Result<(), String> {
    for entry in check!(fs::read_dir(path)) {
        let entry = check!(entry);
        let name: String = entry
            .path()
            .file_name()
            .unwrap()
            .to_string_lossy()
            .into_owned();
        let filePathName: String = format!("{}/{}", path.display(), name);
        let filepath: &Path = &Path::new(&filePathName);
        let rname = rpath.clone() + &name;
        if filepath.is_dir() {
            compile_folder(filepath, rname + ".")?;
        } else if filePathName.ends_with(".lua") {
            let code = compile_file(filepath, name, 2)?;
            let compiled_path = format!(
                "{}{}",
                filepath.display().to_string().strip_suffix(".lua").unwrap(),
                ".clue"
            );
            if !arg!(ENV_DONTSAVE) {
                check!(fs::write(compiled_path, code))
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), String> {
    let cli = Cli::parse();
    if cli.license {
        println!("{}", include_str!("../LICENSE"));
        return Ok(());
    }
    unsafe {
        ENV_TOKENS = cli.tokens;
        ENV_STRUCT = cli.r#struct;
        ENV_OUTPUT = cli.output;
        ENV_DONTSAVE = cli.dontsave;
        ENV_PATHISCODE = cli.pathiscode;
        ENV_NOMULTILINE = cli.nomultiline;
    }
    let codepath = cli.path.unwrap();
    if arg!(ENV_PATHISCODE) {
        println!(
            "{}",
            compile_code(codepath.clone(), "(command line)".to_owned(), 0)?
        );
        return Ok(());
    }
    let path: &Path = Path::new(&codepath);

    if path.is_dir() {
        compile_folder(path, String::new())?;
    } else if path.is_file() {
        let compiled_name =
            String::from(path.display().to_string().strip_suffix(".lua").unwrap()) + ".clue";
        let code = compile_file(
            path,
            path.file_name().unwrap().to_string_lossy().into_owned(),
            0,
        )?;

        if !arg!(ENV_DONTSAVE) {
            check!(fs::write(compiled_name, code))
        }
    } else {
        return Err(String::from("The given path doesn't exist"));
    }
    Ok(())
}

#[cfg(test)]
mod test {
    use crate::compile_folder;
    use std::path::Path;

    #[test]
    fn compile() {
        compile_folder(Path::new("tests/"), String::new()).unwrap();
    }
}
