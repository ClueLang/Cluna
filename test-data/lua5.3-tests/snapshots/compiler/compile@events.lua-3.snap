---
source: src/main.rs
expression: compiled
input_file: test-data/lua5.3-tests/events.lua
---
print('testing metatables');
local debug = require('debug')
X = 20
B = 30
_ENV = setmetatable({}, {
    __index = _G
})
collectgarbage();
X = X + 10
assert(X == 30 && _G.X == 20);
B = false
assert(B == false);
B = nil
assert(B == 30);
assert(getmetatable({}) == nil);
assert(getmetatable(4) == nil);
assert(getmetatable(nil) == nil);
a = {
    name = "NAME"
}
setmetatable(a, {
    __metatable = "xuxu", 
    __tostring = fn (x) {
        return x.name
    }
});
assert(getmetatable(a) == "xuxu");
assert(tostring(a) == "NAME");
assert(pcall(setmetatable, a, {}) == false);
a.name = "gororoba"
assert(tostring(a) == "gororoba");
local a, t = {
    10, 
    20, 
    30, 
    x = "10", 
    y = "20"
}, {}
assert(setmetatable(a, t) == a);
assert(getmetatable(a) == t);
assert(setmetatable(a, nil) == a);
assert(getmetatable(a) == nil);
assert(setmetatable(a, t) == a);
global fn f(t, i, e) {
    assert(!e);
    local p = rawget(t, "parent")
    return (p && p[(i)] + 3), "dummy return"
}
t.__index = f
a.parent = {
    z = 25, 
    x = 12, 
    4 = 24
}
assert(a[(1)] == 10 && a.z == 28 && a[(4)] == 27 && a.x == "10");
collectgarbage();
a = setmetatable({}, t)
global fn f(t, i, v) {
    rawset(t, i, v - 3);
}
setmetatable(t, t);
t.__newindex = f
a[(1)] = 30
a.x = "101"
a[(5)] = 200
assert(a[(1)] == 27 && a.x == 98 && a[(5)] == 197);
{
    local mt = {}
    mt.__newindex = mt
    local t = setmetatable({}, mt)
    t[(1)] = 10
    assert(mt[(1)] == 10);
}
local c = {}
a = setmetatable({}, t)
t.__newindex = c
a[(1)] = 10
a[(2)] = 20
a[(3)] = 90
assert(c[(1)] == 10 && c[(2)] == 20 && c[(3)] == 90);
{
    local a
    a = setmetatable({}, {
        __index = setmetatable({}, {
            __index = setmetatable({}, {
                __index = fn (_, n) {
                    return a[(n - 3)] + 4, "lixo"
                }
            })
        })
    })
    a[(0)] = 20
    for i = 0, 10 {
        assert(a[(i * 3)] == 20 + i * 4);
    }
}
{
    local foi
    local a = {}
    for i = 1, 10 {
        a[(i)] = 0
        a[('a' .. i)] = 0
    }
    setmetatable(a, {
        __newindex = fn (t, k, v) {
            foi = true
            rawset(t, k, v);
        }
    });
    foi = false
    a[(1)] = 0
    assert(!foi);
    foi = false
    a[('a1')] = 0
    assert(!foi);
    foi = false
    a[('a11')] = 0
    assert(foi);
    foi = false
    a[(11)] = 0
    assert(foi);
    foi = false
    a[(1)] = nil
    assert(!foi);
    foi = false
    a[(1)] = nil
    assert(foi);
}
setmetatable(t, nil);
global fn f(t, ...) {
    return t, {
        ...
    }
}
t.__call = f
{
    local x, y = a(table.unpack({
        'a', 
        1
    }))
    assert(x == a && y[(1)] == 'a' && y[(2)] == 1 && y[(3)] == nil);
    x, y = a()
    assert(x == a && y[(1)] == nil);
}
local b = setmetatable({}, t)
setmetatable(b, t);
global fn f(op) {
    return fn (...) {
        cap = {
            0 = op, 
            ...
        }
        return (...)
    }
}
t.__add = f("add")
t.__sub = f("sub")
t.__mul = f("mul")
t.__div = f("div")
t.__idiv = f("idiv")
t.__mod = f("mod")
t.__unm = f("unm")
t.__pow = f("pow")
t.__len = f("len")
t.__band = f("band")
t.__bor = f("bor")
t.__bxor = f("bxor")
t.__shl = f("shl")
t.__shr = f("shr")
t.__bnot = f("bnot")
assert(b + 5 == b);
assert(cap[(0)] == "add" && cap[(1)] == b && cap[(2)] == 5 && cap[(3)] == nil);
assert(b + '5' == b);
assert(cap[(0)] == "add" && cap[(1)] == b && cap[(2)] == '5' && cap[(3)] == nil);
assert(5 + b == 5);
assert(cap[(0)] == "add" && cap[(1)] == 5 && cap[(2)] == b && cap[(3)] == nil);
assert('5' + b == '5');
assert(cap[(0)] == "add" && cap[(1)] == '5' && cap[(2)] == b && cap[(3)] == nil);
b = b - 3
assert(getmetatable(b) == t);
assert(5 - a == 5);
assert(cap[(0)] == "sub" && cap[(1)] == 5 && cap[(2)] == a && cap[(3)] == nil);
assert('5' - a == '5');
assert(cap[(0)] == "sub" && cap[(1)] == '5' && cap[(2)] == a && cap[(3)] == nil);
assert(a * a == a);
assert(cap[(0)] == "mul" && cap[(1)] == a && cap[(2)] == a && cap[(3)] == nil);
assert(a / 0 == a);
assert(cap[(0)] == "div" && cap[(1)] == a && cap[(2)] == 0 && cap[(3)] == nil);
assert(a % 2 == a);
assert(cap[(0)] == "mod" && cap[(1)] == a && cap[(2)] == 2 && cap[(3)] == nil);
assert(a /_ (1 / 0) == a);
assert(cap[(0)] == "idiv" && cap[(1)] == a && cap[(2)] == 1 / 0 && cap[(3)] == nil);
assert(a & "hi" == a);
assert(cap[(0)] == "band" && cap[(1)] == a && cap[(2)] == "hi" && cap[(3)] == nil);
assert(a | "hi" == a);
assert(cap[(0)] == "bor" && cap[(1)] == a && cap[(2)] == "hi" && cap[(3)] == nil);
assert("hi" ^^ a == "hi");
assert(cap[(0)] == "bxor" && cap[(1)] == "hi" && cap[(2)] == a && cap[(3)] == nil);
assert(-a == a);
assert(cap[(0)] == "unm" && cap[(1)] == a);
assert(a ^ 4 == a);
assert(cap[(0)] == "pow" && cap[(1)] == a && cap[(2)] == 4 && cap[(3)] == nil);
assert(a ^ '4' == a);
assert(cap[(0)] == "pow" && cap[(1)] == a && cap[(2)] == '4' && cap[(3)] == nil);
assert(4 ^ a == 4);
assert(cap[(0)] == "pow" && cap[(1)] == 4 && cap[(2)] == a && cap[(3)] == nil);
assert('4' ^ a == '4');
assert(cap[(0)] == "pow" && cap[(1)] == '4' && cap[(2)] == a && cap[(3)] == nil);
assert(#a == a);
assert(cap[(0)] == "len" && cap[(1)] == a);
assert(~a == a);
assert(cap[(0)] == "bnot" && cap[(1)] == a);
assert(a << 3 == a);
assert(cap[(0)] == "shl" && cap[(1)] == a && cap[(2)] == 3);
assert(1.5 >> a == 1.5);
assert(cap[(0)] == "shr" && cap[(1)] == 1.5 && cap[(2)] == a);
t = setmetatable({
    1, 
    2, 
    3
}, {
    __len = fn () {
        return 10
    }
})
assert(#t == 10 && rawlen(t) == 3);
assert(rawlen("abc") == 3);
assert(!pcall(rawlen, io.stdin));
assert(!pcall(rawlen, 34));
assert(!pcall(rawlen));
assert(rawlen(string.rep('a', 1000)) == 1000);
t = {}
t.__lt = fn (a, b, c) {
    collectgarbage();
    assert(c == nil);
    if type(a) == 'table' {
        a = a.x
    }
    if type(b) == 'table' {
        b = b.x
    }
    return a < b, "dummy"
}
global fn Op(x) {
    return setmetatable({
        x = x
    }, t)
}
local fn test() {
    assert(!(Op(1) < Op(1)) && (Op(1) < Op(2)) && !(Op(2) < Op(1)));
    assert(!(1 < Op(1)) && (Op(1) < 2) && !(2 < Op(1)));
    assert(!(Op('a') < Op('a')) && (Op('a') < Op('b')) && !(Op('b') < Op('a')));
    assert(!('a' < Op('a')) && (Op('a') < 'b') && !(Op('b') < Op('a')));
    assert((Op(1) <= Op(1)) && (Op(1) <= Op(2)) && !(Op(2) <= Op(1)));
    assert((Op('a') <= Op('a')) && (Op('a') <= Op('b')) && !(Op('b') <= Op('a')));
    assert(!(Op(1) > Op(1)) && !(Op(1) > Op(2)) && (Op(2) > Op(1)));
    assert(!(Op('a') > Op('a')) && !(Op('a') > Op('b')) && (Op('b') > Op('a')));
    assert((Op(1) >= Op(1)) && !(Op(1) >= Op(2)) && (Op(2) >= Op(1)));
    assert((1 >= Op(1)) && !(1 >= Op(2)) && (Op(2) >= 1));
    assert((Op('a') >= Op('a')) && !(Op('a') >= Op('b')) && (Op('b') >= Op('a')));
    assert(('a' >= Op('a')) && !(Op('a') >= 'b') && (Op('b') >= Op('a')));
}
test();
t.__le = fn (a, b, c) {
    assert(c == nil);
    if type(a) == 'table' {
        a = a.x
    }
    if type(b) == 'table' {
        b = b.x
    }
    return a <= b, "dummy"
}
test();
local fn rawSet(x) {
    local y = {}
    for _, k with pairs(x) {
        y[(k)] = 1
    }
    return y
}
local fn Set(x) {
    return setmetatable(rawSet(x), t)
}
t.__lt = fn (a, b) {
    for k with pairs(a) {
        if !b[(k)] {
            return false
        }
        b[(k)] = nil
    }
    return next(b) != nil
}
t.__le = nil
assert(Set({
    1, 
    2, 
    3
}) < Set({
    1, 
    2, 
    3, 
    4
}));
assert(!(Set({
    1, 
    2, 
    3, 
    4
}) < Set({
    1, 
    2, 
    3, 
    4
})));
assert((Set({
    1, 
    2, 
    3, 
    4
}) <= Set({
    1, 
    2, 
    3, 
    4
})));
assert((Set({
    1, 
    2, 
    3, 
    4
}) >= Set({
    1, 
    2, 
    3, 
    4
})));
assert((Set({
    1, 
    3
}) <= Set({
    3, 
    5
})));
t.__le = fn (a, b) {
    for k with pairs(a) {
        if !b[(k)] {
            return false
        }
    }
    return true
}
assert(!(Set({
    1, 
    3
}) <= Set({
    3, 
    5
})));
assert(!(Set({
    1, 
    3
}) <= Set({
    3, 
    5
})));
assert(!(Set({
    1, 
    3
}) >= Set({
    3, 
    5
})));
t.__eq = fn (a, b) {
    for k with pairs(a) {
        if !b[(k)] {
            return false
        }
        b[(k)] = nil
    }
    return next(b) == nil
}
local s = Set({
    1, 
    3, 
    5
})
assert(s == Set({
    3, 
    5, 
    1
}));
assert(!rawequal(s, Set({
    3, 
    5, 
    1
})));
assert(rawequal(s, s));
assert(Set({
    1, 
    3, 
    5, 
    1
}) == rawSet({
    3, 
    5, 
    1
}));
assert(rawSet({
    1, 
    3, 
    5, 
    1
}) == Set({
    3, 
    5, 
    1
}));
assert(Set({
    1, 
    3, 
    5
}) != Set({
    3, 
    5, 
    1, 
    6
}));
t[(Set({
    1, 
    3, 
    5
}))] = 1
assert(t[(Set({
    1, 
    3, 
    5
}))] == nil);
if !T {
    (Message || print)('\n >>> testC not active: skipping tests for \z
userdata equality <<<\n');
} else {
    local u1 = T.newuserdata(0)
    local u2 = T.newuserdata(0)
    local u3 = T.newuserdata(0)
    assert(u1 != u2 && u1 != u3);
    debug.setuservalue(u1, 1);
    debug.setuservalue(u2, 2);
    debug.setuservalue(u3, 1);
    debug.setmetatable(u1, {
        __eq = fn (a, b) {
            return debug.getuservalue(a) == debug.getuservalue(b)
        }
    });
    debug.setmetatable(u2, {
        __eq = fn (a, b) {
            return true
        }
    });
    assert(u1 == u3 && u3 == u1 && u1 != u2);
    assert(u2 == u1 && u2 == u3 && u3 == u2);
    assert(u2 != {});
}
t.__concat = fn (a, b, c) {
    assert(c == nil);
    if type(a) == 'table' {
        a = a.val
    }
    if type(b) == 'table' {
        b = b.val
    }
    if A {
        return a .. b
    } else {
        return setmetatable({
            val = a .. b
        }, t)
    }
}
c = {
    val = "c"
}
setmetatable(c, t);
d = {
    val = "d"
}
setmetatable(d, t);
A = true
assert(c .. d == 'cd');
assert(0 .. "a" .. "b" .. c .. d .. "e" .. "f" .. (5 + 3) .. "g" == "0abcdef8g");
A = false
assert((c .. d .. c .. d).val == 'cdcd');
x = c .. d
assert(getmetatable(x) == t && x.val == 'cd');
x = 0 .. "a" .. "b" .. c .. d .. "e" .. "f" .. "g"
assert(x.val == "0abcdefg");
c = {}
local x
setmetatable(c, {
    __concat = fn (a, b) {
        assert(type(a) == "number" && b == c || type(b) == "number" && a == c);
        return c
    }
});
assert(c .. 5 == c && 5 .. c == c);
assert(4 .. c .. 5 == c && 4 .. 5 .. 6 .. 7 .. c == c);
local t1, t2, c, d
t1 = {}
c = {}
setmetatable(c, t1);
d = {}
t1.__eq = fn () {
    return true
}
t1.__lt = fn () {
    return true
}
setmetatable(d, t1);
assert(c == d && c < d && !(d <= c));
t2 = {}
t2.__eq = t1.__eq
t2.__lt = t1.__lt
setmetatable(d, t2);
assert(c == d && c < d && !(d <= c));
local i
local tt = {
    __call = fn (t, ...) {
        i = i + 1
        if t.f {
            return t.f(...)
        } else {
            return {
                ...
            }
        }
    }
}
local a = setmetatable({}, tt)
local b = setmetatable({
    f = a
}, tt)
local c = setmetatable({
    f = b
}, tt)
i = 0
x = c(3, 4, 5)
assert(i == 3 && x[(1)] == 3 && x[(3)] == 5);
assert(_G.X == 20);
print('+');
local _g = _G
_ENV = setmetatable({}, {
    __index = fn (_, k) {
        return _g[(k)]
    }
})
a = {}
rawset(a, "x", 1, 2, 3);
assert(a.x == 1 && rawget(a, "x", 3) == 1);
print('+');
mt = {
    __index = fn (a, b) {
        return a + b
    }, 
    __len = fn (x) {
        return math.floor(x)
    }
}
debug.setmetatable(10, mt);
assert(getmetatable(-2) == mt);
assert((10)[(3)] == 13);
assert((10)[("3")] == 13);
assert(#3.45 == 3);
debug.setmetatable(23, nil);
assert(getmetatable(-2) == nil);
debug.setmetatable(true, mt);
assert(getmetatable(false) == mt);
mt.__index = fn (a, b) {
    return a || b
}
assert((true)[(false)] == true);
assert((false)[(false)] == false);
debug.setmetatable(false, nil);
assert(getmetatable(true) == nil);
debug.setmetatable(nil, mt);
assert(getmetatable(nil) == mt);
mt.__add = fn (a, b) {
    return (a || 0) + (b || 0)
}
assert(10 + nil == 10);
assert(nil + 23 == 23);
assert(nil + nil == 0);
debug.setmetatable(nil, nil);
assert(getmetatable(nil) == nil);
debug.setmetatable(nil, {});
a = {}
setmetatable(a, a);
a.__index = a
a.__newindex = a
assert(!pcall(fn (a, b) {
    return a[(b)]
}, a, 10));
assert(!pcall(fn (a, b, c) {
    a[(b)] = c
}, a, 10, true));
T, K, V = nil
grandparent = {}
grandparent.__newindex = fn (t, k, v) {
    T = t
    K = k
    V = v
}
parent = {}
parent.__newindex = parent
setmetatable(parent, grandparent);
child = setmetatable({}, parent)
child.foo = 10
assert(T == parent && K == "foo" && V == 10);
print('OK');
return 12
